cpp/src/random_projection/rproj_utils.cuh

```diff
@@ -25,14 +25,13 @@
 const int TPB_X = 256;
 
 inline void sample_without_replacement(size_t n_population, size_t n_samples,
-                                       int* indices, size_t& indices_idx) {
-  std::random_device dev;
-  std::mt19937 gen(dev());
-
+                                       int* indices, size_t offset,
+                                       int random_state) {
+  std::mt19937 gen(random_state);
   std::uniform_int_distribution<int> uni_dist(0, n_population - 1);
 
+  int indices_idx = offset;
   std::unordered_set<int> s;
-
   for (size_t i = 0; i < n_samples; i++) {
```

Comment:
This is pretty expensive here. It would be nice to see this placed on device at some point (not necessary for this PR).